### PR TITLE
AssertExp: make it a bit more formal and correct

### PR DESCRIPTION
**Pull request by WalterBright**
Correct omissions in the original, and clearly mark implementation defined behavior, undefined behavior, and best practices while avoiding compiler-specific details.

Relies on new macros in https://github.com/dlang/dlang.org/pull/2091

### CODE REVIEW

File: spec/expression.dd [hunk 1]
```diff
@@ -1671,19 +1671,21 @@ $(GNAME AssertExpression):
...
-    $(P The expression $(D assert(0)) is a special case; it
-        signifies that it is unreachable code.
-        Without optimizations, an $(D AssertError) is thrown at runtime.
-        In release mode ($(LINK2 $(ROOT_DIR)/dmd.html#switch-release, `-release`))
-        the execution is halted
-        (on the x86 processor, a $(D HLT) instruction can be used to halt
-        execution).
-        The optimization and code generation phases of compilation may
-        assume that it is unreachable code.
+    The implementation may handle $(D assert(0))  differently in that in release mode
```
Comment (wilzbach) [inline spec/expression.dd]: double space

File: spec/expression.dd [hunk 2]
```diff
@@ -1623,44 +1623,44 @@ $(GNAME AssertExpression):
...
+    any side effects specified by the $(I AssertExpression) may not occur.
+    The behavior if the first $(I AssertExpression) is evaluated and is false
+    is also typically set with a compiler switch and may include these options:
+    $(OL
+	$(LI continuing execution)
+	$(LI immediately halting via execution of a special CPU instruction)
+	$(LI aborting the program)
+	$(LI calling the assert failure function in the corresponding C
+	runtime library)
+	$(LI throwing the $(D AssertError) exception in the D runtime library)
```
Comment (wilzbach) [inline spec/expression.dd]: Don't use tabs.

**Comment by wilzbach** [comment]
DAutoTest was failing due to:

```
> Searching for undefined macros
/dev/shm/dtest/work/repo/dlang.org/web/dlangspec.html:7265:    <!--UNDEFINED MACRO: "UNDEFINED_BEHAVIOR"-->
/dev/shm/dtest/work/repo/dlang.org/web/dlangspec.html:7267:    <!--UNDEFINED MACRO: "IMPLEMENTATION_DEFINED"-->
/dev/shm/dtest/work/repo/dlang.org/web/dlangspec.html:7269:    <!--UNDEFINED MACRO: "BEST_PRACTICE"-->
/dev/shm/dtest/work/repo/dlang.org/web/spec/expression.html:1752:    <!--UNDEFINED MACRO: "UNDEFINED_BEHAVIOR"-->
/dev/shm/dtest/work/repo/dlang.org/web/spec/expression.html:1754:    <!--UNDEFINED MACRO: "IMPLEMENTATION_DEFINED"-->
/dev/shm/dtest/work/repo/dlang.org/web/spec/expression.html:1756:    <!--UNDEFINED MACRO: "BEST_PRACTICE"-->
```

I rebased your PR to rekick DAutoTest -> should be passing now.

File: spec/expression.dd [hunk 3]
```diff
@@ -1623,44 +1623,44 @@ $(GNAME AssertExpression):
...
-        debugging tool and as a way of communicating to the compiler facts
-        about the code that it may employ to produce more efficient code.
+    $(P The first $(I AssignExpression) must evaluate to true. If it does not, an $(I Assert Failure)
+    has occurred and the program enters an $(I Invalid State).
     )
 
-    $(P Programs for which $(I AssignExpression) is false are invalid.
-        Subsequent to such a false result, the program is in an invalid,
-        non-recoverable state.
+    $(P The expression $(D assert(0)) is a special case; it
```
Comment (andralex) [inline spec/expression.dd]: We need to clarify this further. Are these also special-cased?

```d
assert(1 - 1);
assert(10 * 10 - 100 == 0);
enum a = 0;
assert(a);
enum b = sqrt(100);
assert(b == 10);
assert(false);
assert(null);
assert([]);
```

In other words, to what extent does the compiler go to evaluate the expression during compilation? I don't know the answer.

File: spec/expression.dd [hunk 4]
```diff
@@ -1623,44 +1623,44 @@ $(GNAME AssertExpression):
...
+    signifies that it is unreachable code.
```
Comment (andralex) [inline spec/expression.dd]: This paragraph should also discuss `assert(0, message);` If I remember correctly, the message is displayed in debug mode and not in release mode (which btw is irritating).

**Comment by WalterBright** [comment]
> different semantics inside unittest blocks and contracts, but it's not defined exactly what these different semantics are.

Right, and I plan to address that in the unittest and contract sections, which will be rewritten as well.

File: spec/expression.dd [hunk 5]
```diff
@@ -1671,19 +1672,21 @@ $(GNAME AssertExpression):
...
     $(P When compiled and run, it will produce the message:)
 
-$(CONSOLE
-core.exception.AssertError@test.d(3) an error message
-)
+    $(CONSOLE core.exception.AssertError@test.d(3) an error message)
+
+    The implementation may handle $(D assert(0)) differently in that in release mode
+    it may simply generate a $(D HLT) instruction or equivalent, even
+    in release mode.
```
Comment (aG0aep6G) [inline spec/expression.dd]: Repetition of "in release mode".

Comment (andralex) [reply spec/expression.dd]: @aG0aep6G you may want to make obvious/noncontroversial fixes like this by editing the PR and committing to it directly. It works real nice... props for the github team.

**Comment by andralex** [comment]
The phrases italicized and title cased, as in $(I Assert Failure), would be only meaningful if defined elsewhere.

File: spec/expression.dd (hunk 5 above)
Comment (aG0aep6G) [reply spec/expression.dd]: > @aG0aep6G you may want to make obvious/noncontroversial fixes like this by editing the PR and committing to it directly. It works real nice... props for the github team.

I don't think I have the powers to do that. I'm not a member of Team D or anything.

Comment (andralex) [reply spec/expression.dd]: @aG0aep6G if you go to "files", do you see something like a crayon in the decoration at the top of each file changed?

Comment (aG0aep6G) [reply spec/expression.dd]: > @aG0aep6G if you go to "files", do you see something like a crayon in the decoration at the top of each file changed?

Yeah. It's not clickable and the tooltip says: "You must be signed in and have push access to the AssertExp branch to make changes." I don't have that kind of access to the D repositories.

**Comment by quickfur** [comment]
@andralex For obvious reasons, the PR editing function is only available to people who have commit access to the repo (other than the PR submitter himself, of course).

**Comment by WalterBright** [comment]
@andralex I made some changes to address the comments.(but not the italicized one)

**Comment by andralex** [comment]
My changes have been overwritten, sigh. @WalterBright please avoid force pushing...

**Comment by WalterBright** [comment]
Generally incomprehensible errors and warnings from DAutoTest.

File: spec/expression.dd [hunk 6]
```diff
@@ -1623,44 +1623,54 @@ $(GNAME AssertExpression):
...
+    which a $(I Struct Invariant) exists, the $(I Struct Invariant) must hold.
+    )
+
+    $(P The type of an $(I AssertExpression) is $(D void).
+    )
+
+    $(UNDEFINED_BEHAVIOR Once in an $(I Invalid State) the behavior of the continuing execution
+    of the program is undefined.)
+
+    $(IMPLEMENTATION_DEFINED Whether the first $(I AssertExpression) is evaluated
```
Comment (wilzbach) [inline spec/expression.dd]: You never close this ... any decent macro language would have made it _a lot_ easier to spot the error ;-)
Also your PR needs to be rebased against `upstream/master`, s.t. it includes the new macros.

Comment (WalterBright) [reply spec/expression.dd]: thanks!

**Pull request merged**